### PR TITLE
Add SudoFile function and use it with root file system access is needed.

### PR DESCRIPTION
**Pull request by john-gress**
See https://rally1.rallydev.com/#/detail/userstory/98723196460?fdp=true

See related pull requests:
https://github.com/LogRhythm/QueueNado/pull/46
https://github.schq.secious.com/Logrhythm/NMSetup/pull/27
https://github.schq.secious.com/Logrhythm/Research/pull/1720

### CODE REVIEW

File: test/ToolsTestFileIO.cpp [hunk 1]
```diff
@@ -642,15 +643,48 @@ TEST_F(TestFileIO, TestReadAsciiFileContentAsRoot) {
...
 
+TEST_F(TestFileIO, TestSudoFileRemoveFile) {
+
+   // Start as root user
+   int previousUID = setfsuid(-1);
+   int previousGID = setfsgid(-1);
+   ASSERT_EQ(previousUID, 0);
+   ASSERT_EQ(previousGID, 0);
+
+   std::string filename{"/tmp/TestFileIO_"};
```
Comment (alexweltman) [inline test/ToolsTestFileIO.cpp]: seems weird to name the file something random, instead of putting the files in a directory for this test

**Comment by alexweltman** [comment]
👍

File: src/FileIO.h [hunk 2]
```diff
@@ -60,4 +61,26 @@ struct ScopedFileDescriptor {
       close(fd);
    }
 };
+
+// Do file access as root user for the function passed in.
+// Example: auto result = FileIO::SudoFile(FileIO::ReadAsciiFileContent, filePath);
```
Comment (KjellKod) [inline src/FileIO.h]: I'm missing something obvious. What is the reasoning with the "File" ending? Wouldn't "SudoAccess" or just "Sudo" be more appropriate?

File: src/FileIO.h [hunk 3]
```diff
@@ -60,4 +61,26 @@ struct ScopedFileDescriptor {
...
+template<typename FnCall, typename... Args>
+auto SudoFile(FnCall fn, Args&& ... args) {
+   static std::mutex mPermissionsMutex2;
```
Comment (KjellKod) [inline src/FileIO.h]: The "2" can now be removed

File: src/FileIO.h (hunk 2 above)
Comment (john-gress) [inline src/FileIO.h]: "sudo" or superuser do gives the user full superuser access. Theoretically we are only elevate file system access to have root privileges, hence SudoFile. That said, I'm not opposed to changing the name. Let me know what you prefer.

File: src/FileIO.h (hunk 3 above)
Comment (john-gress) [inline src/FileIO.h]: Changed from mPermissionsMutex2 to sudoFileMutex.

File: test/ToolsTestFileIO.cpp (hunk 1 above)
Comment (john-gress) [inline test/ToolsTestFileIO.cpp]: I agree. I changed this to use the mTestDirectory which is already setup and cleaned up.

File: src/FileIO.h (hunk 2 above)
Comment (KjellKod) [inline src/FileIO.h]: I don't want to go into nitpicking mode so I think it's fine. I'll wait to plus it till I've tested it but review wise it's fine 

(If Alex has tested it already then I'm good with merging)

File: src/FileIO.h [hunk 4]
```diff
@@ -60,4 +61,26 @@ struct ScopedFileDescriptor {
       close(fd);
    }
 };
+
+// Do file access as root user for the function passed in.
+// Example: auto result = FileIO::SudoFile(FileIO::ReadAsciiFileContent, filePath);
+template<typename FnCall, typename... Args>
+auto SudoFile(FnCall fn, Args&& ... args) {
```
Comment (KjellKod) [inline src/FileIO.h]: i.e. it *also* shows that the template substitution failed but it's way too much information. Unless you are skilled with templates the first error message is much easier to understand.

Comment (john-gress) [inline src/FileIO.h]: Good call. I'll add the return type.

**Comment by KjellKod** [comment]
👍

**Pull request merged**